Flooring rewrite.

MistakeNot4892 opened the pull request:
http://i.imgur.com/fYsBy7T.png
http://i.imgur.com/v7Ijmyz.png
http://i.imgur.com/GU7PBHy.png

The floor icon system has been stripped back and rewritten to unify construction, removal and sprite generation of tiles. Ideally very little of this should be visible to the average player, but there have been several major results of the project which I will detail below.

Purpose
- Reduce snowflake/boilerplate code for smoothed turfs (carpet, grass siding, etc).
- Clean up clutter in the /turf/simulated/floor tree.
- Make tiles and floor coverings more intuitive and properly structured.
- Move turf details like signs or coloured tiles to overlays, reducing map complexity, icon count and margin for error.
- Force coders/mappers to properly define flooring in a centralized location instead of using instances.
- Prevent coders/mappers from using excessive nonstandard instances for cosmetic details (and thereby minimizing map load fails down the line, looking at you Luna).
- Convert some of the snowflake bullshit from turfs to machinery - specifically mech rechargers and light tiles.

Outcomes:
- /turf/simulated/floor is now the basic plating turf, /turf/simulated/floor/tiled is the 'standard' station turf.
- The flooring var on the above determines sprite, description, tile behavior, etc. using the /flooring/ datum.
- /obj/effect/floor_decal now handles mapping in things like signs or coloured edging. The obj qdels() itself and adds its icon to the turf overlays in initialize().
- The map is currently very very barebones in regards to cosmetic detail - all of the department-specific siding and a lot of the niche flooring types will need to be readded/replaced. This is very easy but pretty time consuming.

Todo:
- [x] Replace the map cosmetic detail/decals.
- [x] Point all the unsimulated/shuttle turfs at the correct icon files.
- [x] Add sprites/flooring definitions/prefabs for dark tiles, white tiles, lino and the freezer.
- [x] Update the holodeck maps.
- [x] Remap the atmospherics over-turf piping.
- [x] Fix up/test asteroid overlays.
- [x] Reimplement light tiles.
- [x] Update/reimplement floor painter.
- [X] Reimplement mech rechargers.
- [x] Triple-check the various air turfs on the map (engine etc).
- [x] Fix pipe layering over turf and qdel when wrenched on a turf with flooring.
- [X] Drink heavily.

Kelenius committed:
Mech chargers are now a single machine

Can be built and upgraded
Resolves #5215

MistakeNot4892 committed:
Removed turf_animation for now, ported vg hasproximity optimisation. Rewrite and refactor of how turfs handle icon updates, edge smoothing and
construction/deconstruction. Major map rework/changes.

mwerezak commented on code/game/turfs/flooring/flooring.dm:
```diff
@@ -0,0 +1,133 @@
+	var/name
+	var/desc
+	var/icon
+	var/icon_base
```
Name might confusing, any reason to not just call it icon_state?

Mloc replied:
really nitpicky but I've been using stuff prefixed by /decl instead of root things like this to keep / from being cluttered

mwerezak replied:
Does that make a difference for anything other than DM's object browser?

Mloc replied:
nope! you can't even define procs/vars on /, it's just nitpickery

mwerezak commented on code/game/turfs/flooring/flooring.dm:
```diff
@@ -0,0 +1,133 @@
+				del(F)
+				continue
+			flooring_types[F.name] = F
+	return flooring_types[flooring_name]
```
Indexing by flooring name seems kind of iffy. It might be worth indexing by the type path itself.

Zuhayr replied:
This will make it a pain to look up types of flooring via the proc.

mwerezak replied:
I'm not sure what you mean.

Zuhayr replied:
If they are indexed by type path then I'm going to have to supply the type path to get_flooring_by_name(). I guess that's not a huge issue.

mwerezak replied:
Yeah, that's the idea. You supply the type to obtain the reference to the flooring datum. It seems more robust than supplying the name since the type uniquely identifies the type of flooring, while name does not necessarily do so.

mwerezak replied:
It will also allow you to populate the `flooring_types` list lazily, creating new flooring datum instances only when they are requested and not yet created.

mwerezak replied:
This proc would then simplify to:
```dm
/proc/get_flooring_instance(var/floortype)
    if(!(floortype in flooring_types))
        flooring_types[floortype] = new floortype
    return flooring_types[floortype]
```

mwerezak commented on code/game/turfs/flooring/flooring_premade.dm:
```diff
@@ -0,0 +1,107 @@
+	name = "carpet"
+	icon = 'icons/turf/flooring/carpet.dmi'
+	icon_state = "carpet"
+	initial_flooring = "carpet"
```
If the flooring is loaded when the turf is created then we can just get rid of most of these initializations. Less things to maintain.

Zuhayr replied:
Those lines are for the benefit of the map.

mwerezak replied:
Ah, right.

mwerezak commented on code/game/turfs/turf.dm:
```diff
@@ -105,48 +100,38 @@
-				O.HasProximity(A, 1)
-			return
+	if(A && (A.flags & PROXMOVE))
+		for(var/atom/thing as mob|obj|turf|area in range(1))
```
I wonder if this is better placed in the specific atom's Move() instead of in turf code.

```
/obj/proxthing/Move()
    if(..())
        //do stuff
```

Zuhayr replied:
Might be, but that's a little outside of scope tbh.

mwerezak replied:
Your call, but that way turfs don't have to check every atom that enters them for a PROXMOVE flag. Instead the atoms in question would check if they've succesfully moved (by the return value from `..()`), and instead of being a hardcoded call to `thing.HasProximity(A, 1)`, they could do whatever they want.

atlantiscze commented:
Regarding mech rechargers, i believe PR is up that already unified them into single machine, so i guess it saves you part of the work (#10190).

scrdest commented:
I can't find it offhand, but tg had a refactor a while back that made turfs store a 'what am I built on?' var so that explosions and such on turfs built on, for example, asteroid turfs turned the ground into asteroid tiles, not space. Consider integration?

PsiOmegaDelta commented:
While it can perhaps be remade to suit this PR better Bay does already have a system for that.

MistakeNot4892 committed:
Changes flooring to use /decl/flooring

MistakeNot4892 committed:
Readds preview icons to floors.dmi for use by unsim turfs.

MistakeNot4892 committed:
Updates instances/tiles for Centcomm.

MistakeNot4892 committed:
Blocked out the holodeck tiles.

mwerezak commented on code/game/mecha/mech_bay.dm:
```diff
@@ -1,218 +1,94 @@
-/turf/simulated/floor/mech_bay_recharge_floor
-	name = "Mech Bay Recharge Station"
+/obj/machinery/mech_recharger
```
Reposting my comment on the original PR, this object should probably have `layer = TURF_LAYER + 0.1`

Zuhayr replied:
It has been editied, you just linecommented too high up for 'outdated diff' to grab your comment.

mwerezak commented on code/game/turfs/simulated/floor_attackby.dm:
```diff
@@ -0,0 +1,90 @@
+			if(flooring || !S || !user || !use_flooring)
+				return
+			if(S.use(use_flooring.build_cost))
+				set_flooring(use_flooring.name)
```
Should be `use_flooring.type`

mwerezak commented on code/game/turfs/flooring/flooring.dm:
```diff
@@ -0,0 +1,154 @@
+	flags = TURF_ACID_IMMUNE | TURF_CAN_BREAK
+
+/decl/flooring/reinforced/circuit/green
+	name = "green processing strata"
```
It might be better to just call these both "processing strata". The colour can be seen visually and it's less clutter for the player.

mwerezak replied:
Guh, looks like these aren't used on anything player facing at the moment. That said...

mwerezak commented on code/game/turfs/simulated/floor.dm:
```diff
@@ -1,604 +1,82 @@
+/turf/simulated/floor/proc/set_flooring(var/flooring_id)
+	make_plating(defer_icon_update = 1)
+	if(flooring_id)
+		flooring = get_flooring_by_name(flooring_id)
```
It might be a good idea to set the name of the turf based on it's new flooring type here. That way whenever the flooring type changes we can ensure that it has the correct name.

Zuhayr replied:
turf/simulated/floor/update_icon, line 10

mwerezak commented:
We may want to put a hold on any PRs with mapping changes until this is merged, depending on how bad you think the conflicts could be.

GinjaNinja32 commented on code/game/turfs/simulated/floor.dm:
```diff
@@ -1,604 +1,82 @@
+
+	//TEMPORARY DEBUGGING LINES, REMOVE LATER
+	if(!flooring && icon_state != "plating")
+		name = "OLDCODE FLOORING WOOP WOOP"
```
I love it.

mwerezak committed:
Adds more flooring decal types

mwerezak commented:
@Zuhayr https://github.com/Zuhayr/Baystation12/pull/13

mwerezak committed:
Maps cosmetic flooring to engineering and most of medical

Zuhayr commented:
Fixing mapchanges won't be a big deal unless they are particularly sizable. Going to piss off a lot of people in the freeze if this sneaks in before the older map projects do.

PsiOmegaDelta committed:
Additional floor decal icons.

Adds grey, purple, and white floor decals.
Adds full red and yellow floors, both for floor and floor tile.
Makes the U-shaped warning sprites match the straight and L-shaped sprites.

PsiOmegaDelta committed:
Adds floor decals to the public Exodus areas.

PsiOmegaDelta committed:
More floor decals.

Adds additional red decal icons.
Adds grey outline decal.

PsiOmegaDelta committed:
Adds floor decals to the security department.

PsiOmegaDelta commented:
![Non-rotated decals](https://www.dropbox.com/s/62p6x6cuh25n9s6/DecalRotation.png?dl=1)
**Bug noted:** The floor decal dir is not respected once IG.

PsiOmegaDelta committed:
Fixes floor decal dir and layer not being respected.

PsiOmegaDelta committed:
Adds (dirty) steel tiles.

PsiOmegaDelta committed:
Adds floor decals to the cargo department.

PsiOmegaDelta committed:
Adds additional floor icons.

Adds additional grey and lime cornel decal icons.
Adds directional Space Station 13 icons.

PsiOmegaDelta committed:
Adds more floor decals to the station.

Decals the maintenance tunnels, research, virology, central primary hallway, and some missed spots.

PsiOmegaDelta committed:
Decals Z-level 3, telecoms.

PsiOmegaDelta committed:
Decals Z-level 4, construction.

PsiOmegaDelta committed:
Adds dusty warning stripes.

PsiOmegaDelta committed:
Decals Z-level 5, asteroid.

PsiOmegaDelta committed:
Decals Z-level 2, Crescent.

MistakeNot4892 committed:
Tidied up after a merge.

MistakeNot4892 committed:
Various small tweaks for the flooring update.

mwerezak committed:
Mapping - minor airlock and other details

Finishes main medical decals (adds light blue and light yellow corner
colours), adds various touch ups to mining and incinerator airlock,
science and security blast doors. Makes the science bomb test range
airless. Removes SS13 placard. Adds privacy curtains to the medical ward
and dorms.

mwerezak committed:
Fixes turf decal overlays layering with holopads

Holopads layer at TURF_LAYER + 0.1 so there is the possibility that turf decals might layer over them. This ensures that decals layer under holopads.

mwerezak committed:
Fixes floorbots, floor attackby flooring setting, and adjusts a few flooring names.

Made set_flooring take a flooring instance, in case calling code already
had a flooring instance it seemed wasteful to have to look up the
flooring all over again.

mwerezak commented:
@Zuhayr https://github.com/Zuhayr/Baystation12/pull/17

PsiOmegaDelta committed:
Holo-decals and fixes.

Adds more holodeck decals.
Allows overlays to be glued unto unsimulated turfs.
Ensures holo-stools use an existing icon state.

PsiOmegaDelta committed:
Fewer decal icon states, shower curtains.

Someone had added shower curtains to the map. Creates proper definitions to avoid colors being lost on cleaning.
Floor corner decals now all utilize a white base state and simply apply a color.

PsiOmegaDelta committed:
Cleans up decal corner icon states, ruined by "Generate Instances from Direction" and replaces shower curtains.

PsiOmegaDelta committed:
Map decaling.

Fixes a number of turfs missing a proper icon state.
Sorts out the holodecks.

MistakeNot4892 committed:
Semi-fixes turf layering issues with pipes and newflooring.

Zuhayr commented:
If anyone has any input on fixing the pipe layering issue, please feel free to chip in, pipe code gave me a nasty headache.

Kelenius committed:
Fixes #10599

Kelenius committed:
Fixes #9332

Kelenius committed:
Reverts #10649, fixes #10621

vol-test committed:
Fixes #10508

mwerezak committed:
Fixes pipes disappearing when wrenching fails

Zuhayr commented:
Had to rebase because of git stupidity. Hopefully haven't introduced any bugs. Major issues are now:
- Pipes are not respecting their layering when created. If a visible pipe is spawned or a pipe is wrenched down on top of flooring, it will become invisible until the tile is ripped p. It will stay visible after the tile is put down again.
- The map needs another pass since there were a few conflicts and I may have missed oldcode flooring.

Zuhayr commented:
This is ready for review.

MistakeNot4892 committed:
Recommitting maps after rebase with a lot of conflicts.

MistakeNot4892 committed:
Compile fixes.

MistakeNot4892 committed:
Merge

MistakeNot4892 committed:
Removed intact var in favour of is_plating() proc, fixed issues with invisible initialized pipes.

MistakeNot4892 committed:
Added proper gas turfs for atmos, remapped atmos turfs.

MistakeNot4892 committed:
Reimplemented floor lights as a machine built at the autolathe.

MistakeNot4892 committed:
Changelog.

MistakeNot4892 committed:
Rewrites the floor painter.

GinjaNinja32 commented on code/game/turfs/simulated/floor.dm:
```diff
@@ -1,604 +1,76 @@
+		set_flooring(get_flooring_data(floortype))
+
+	//TEMPORARY DEBUGGING LINES, REMOVE LATER
+	if(!flooring && icon_state != "plating")
```
.

Zuhayr replied:
<Zuhayr> GinjaNinja32: was going to leave those lines in until it's guaranteed that the map is up to date
<Zuhayr> since there may be several merges between now and then

GinjaNinja32 commented on code/game/turfs/turf.dm:
```diff
@@ -105,48 +107,38 @@
+			spawn(0)
+				thing.HasProximity(A, 1)
+				if ((thing && A) && (thing.flags & PROXMOVE))
+					thing.HasProximity(A, 1)
```
double HasProximity?

Zuhayr replied:
Ah I may have meant A.HasProximity(thing,1)

GinjaNinja32 commented on code/game/turfs/turf_changing.dm:
```diff
@@ -0,0 +1,95 @@
+			fire = old_fire
+
+		if (istype(W,/turf/simulated/floor))
+			W.RemoveLattice()
```
this will always be true, W should be `var/turf/simulated/floor/W`:
```DM
	if(ispath(N, /turf/simulated/floor))
		var/turf/simulated/W = new N( locate(src.x, src.y, src.z) )
		// [...]
		if (istype(W,/turf/simulated/floor))
			W.RemoveLattice()
```

GinjaNinja32 commented:
:+1:, only minor problems. With how much this is gonna conflict every time someone changes the map, better to fix those post-merge.

GinjaNinja32 commented:
re-:+1: as above

GinjaNinja32 commented:
re-:+1: with Travis.

MistakeNot4892 committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into turfs

MistakeNot4892 committed:
Minor cleanup of proxmove, changeturf, oldcode floors.

MistakeNot4892 committed:
Added more colour defines, added functionality to the floor painter.

MistakeNot4892 committed:
Floor instance cleanup.

MistakeNot4892 committed:
Removing unneeded defines.
